Return clusterErr rather than err

FengyunPan opened the pull request:
The updateClusterIngressUIDToMasters() should return clusterErr, not err.
If the 'err' of 'masterCluster, masterUID, err := ic.getMasterCluster()' is nil and the 'clusterErr' of 'clusterObj, clusterErr := api.Scheme.DeepCopy(cluster)' is not nil, updateClusterIngressUIDToMasters() will return ("", nil).

And do not log fallbackUID when fallbackUID is nil.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44810)
<!-- Reviewable:end -->

FengyunPan commented:
/release-note-none

cblecker commented:
@k8s-bot ok to test

FengyunPan commented:
@madhusudancs 
@shashidharatd 
@quinton-hoole 
Can you review this patch? Thank you.

FengyunPan commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

quinton-hoole commented on federation/pkg/federation-controller/ingress/ingress_controller.go:
```diff
@@ -647,13 +648,15 @@ func (ic *IngressController) getMasterCluster() (master *federationapi.Cluster,
 	if clusterErr != nil || !ok {
-		glog.Errorf("Internal error: Failed clone cluster resource while attempting to add master ingress UID annotation (%q = %q) from master cluster %q to cluster %q, will try again later: %v", uidAnnotationKey, masterUID, masterCluster.Name, cluster.Name, err)
-		return "", err
+		errmsg := fmt.Sprintf("Internal error: Failed clone cluster %q resource: %v", cluster.Name, clusterErr)
```
nit: Why 'resource' and not 'error'?  And why remove the detail about UID etc form the log message.  It's useful for debugging, right?

FengyunPan replied:
Yes, you are right, it's useful for debugging. And I am so sorry for my patch's description and removing detail log. I will update it.
I just think the following situation will be bug.
If the 'err' of 'masterCluster, masterUID, err := ic.getMasterCluster()' is nil and the 'clusterErr' of 'clusterObj, clusterErr := api.Scheme.DeepCopy(cluster)' is not nil, updateClusterIngressUIDToMasters() will return ("", nil).

quinton-hoole commented on federation/pkg/federation-controller/ingress/ingress_controller.go:
```diff
@@ -647,13 +648,15 @@ func (ic *IngressController) getMasterCluster() (master *federationapi.Cluster,
-		glog.Errorf("Internal error: Failed clone cluster resource while attempting to add master ingress UID annotation (%q = %q) from master cluster %q to cluster %q, will try again later: %v", uidAnnotationKey, masterUID, masterCluster.Name, cluster.Name, err)
-		return "", err
+		errmsg := fmt.Sprintf("Internal error: Failed clone cluster %q resource: %v", cluster.Name, clusterErr)
+		glog.Errorf(errmsg)
```
It's not clear to me that it's a good idea to both log the error and return it (although I realize that the original code did).  Surely the caller will log the error if you return it here?  It seems like the correct general strategy is to either return the error, or log it, not both?

FengyunPan replied:
I keep them and just return correct "clusterErr".

FengyunPan commented:
@k8s-bot unit test this
@k8s-bot bazel test this

Commit:
Return clusterErr rather than err

The updateClusterIngressUIDToMasters() should return clusterErr,
not err. And do not log fallbackUID when fallbackUID is nil.

quinton-hoole commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44810)*
<!-- Sent from Reviewable.io -->

quinton-hoole commented:
/lgtm
/approve

FengyunPan commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

cblecker commented:
/test pull-kubernetes-federation-e2e-gce

quinton-hoole commented:
/approve no-issue

FengyunPan commented:
/test pull-kubernetes-e2e-kops-aws

fejta commented:
/test pull-kubernetes-e2e-kops-aws
